flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FinishedOperatorSubtaskState.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+/**
+ * A specialized {@link OperatorSubtaskState} used to mark the finished subtasks in the snapshot of
+ * this operator.
+ */
+public class FinishedOperatorSubtaskState extends OperatorSubtaskState {
```

Comment:
Do we need a separate class? Could we just add a flag to the current class?

Reply:
It would be also ok to directly add a flag and introduce a special static `FINISHED` instance. Users could not add states to this instance since the underlying `EmptyList` does not support `add` operator. This should be simpler and I updated with this option~